flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/KeyAndValueSerializer.java

```diff
@@ -0,0 +1,181 @@
+	}
+
+	@Override
+	public Tuple2<byte[], StreamRecord<IN>> copy(
+			Tuple2<byte[], StreamRecord<IN>> from,
+			Tuple2<byte[], StreamRecord<IN>> reuse) {
+		StreamRecord<IN> fromRecord = from.f1;
+		StreamRecord<IN> reuseRecord = reuse.f1;
+
+		IN valueCopy = valueSerializer.copy(fromRecord.getValue(), reuseRecord.getValue());
+		fromRecord.copyTo(valueCopy, reuseRecord);
+		reuse.f0 = Arrays.copyOf(from.f0, from.f0.length);
+		reuse.f1 = reuseRecord;
+		return reuse;
+	}
+
+	@Override
+	public int getLength() {
+		return -1;
+	}
+
+	@Override
+	public void serialize(Tuple2<byte[], StreamRecord<IN>> record, DataOutputView target) throws IOException {
+		if (serializedKeyLength < 0) {
+			target.writeInt(record.f0.length);
+		}
+		target.write(record.f0);
+		StreamRecord<IN> toSerialize = record.f1;
+		target.writeLong(toSerialize.getTimestamp());
+		valueSerializer.serialize(toSerialize.getValue(), target);
+	}
+
+	@Override
+	public Tuple2<byte[], StreamRecord<IN>> deserialize(DataInputView source) throws IOException {
+		final int length = getKeyLength(source);
+		byte[] bytes = new byte[length];
+		source.read(bytes);
+		long timestamp = source.readLong();
+		IN value = valueSerializer.deserialize(source);
+		return Tuple2.of(
+			bytes,
+			new StreamRecord<>(value, timestamp)
+		);
+	}
+
+	@Override
+	public Tuple2<byte[], StreamRecord<IN>> deserialize(Tuple2<byte[], StreamRecord<IN>> reuse, DataInputView source) throws IOException {
+		final int length = getKeyLength(source);
+		byte[] bytes = new byte[length];
+		source.read(bytes);
+		long timestamp = source.readLong();
+		IN value = valueSerializer.deserialize(source);
+		StreamRecord<IN> reuseRecord = reuse.f1;
+		reuseRecord.replace(value, timestamp);
+		reuse.f0 = bytes;
+		reuse.f1 = reuseRecord;
+		return reuse;
+	}
+
+	private int getKeyLength(DataInputView source) throws IOException {
+		final int length;
+		if (serializedKeyLength < 0) {
+			length = source.readInt();
+		} else {
+			length = serializedKeyLength;
+		}
+		return length;
+	}
+
+	@Override
+	public void copy(DataInputView source, DataOutputView target) throws IOException {
+		final int length;
+		if (serializedKeyLength < 0) {
+			length = source.readInt();
+			target.writeInt(length);
+		} else {
+			length = serializedKeyLength;
+		}
+		for (int i = 0; i < length; i++) {
+			target.writeByte(source.readByte());
+		}
+		target.writeLong(source.readLong());
+		valueSerializer.copy(source, target);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		KeyAndValueSerializer<?> that = (KeyAndValueSerializer<?>) o;
+		return Objects.equals(valueSerializer, that.valueSerializer);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(valueSerializer);
+	}
+
+	@Override
+	public TypeSerializerSnapshot<Tuple2<byte[], StreamRecord<IN>>> snapshotConfiguration() {
+		return null;
```

Comment:
Should throw a meaningful exception because it's not meant to be used in a context where snapshots would be needed.